### PR TITLE
Add Utf8String type

### PR DESCRIPTION
**Pull request by GrabYourPitchforks**
This PR contains the type itself and enlightens the VM, the GC, reflection, and certain APIs (`Memory` / `Span`) about the new type. Functionality like culture-aware or case-insensitive comparisons are not yet hooked up. Those are implemented in the feature branch, but I didn't include them here because I'm still working on increasing the test coverage.

All `Utf8String`-related code is currently surrounded by `#ifdef` statements to allow easy removal from release branches.

### CODE REVIEW

**Comment by GrabYourPitchforks** [comment]
Experimental package, ref asm, and unit tests at https://github.com/dotnet/corefx/pull/35989.

File: src/System.Private.CoreLib/shared/System/Text/Unicode/Utf8.cs [hunk 1]
```diff
@@ -0,0 +1,186 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System.Buffers;
+using System.Diagnostics;
+
+namespace System.Text.Unicode
+{
+    public static class Utf8
```
Comment (jkotas) [inline src/System.Private.CoreLib/shared/System/Text/Unicode/Utf8.cs]: This is adding new public APIs. Do we have an approved API review issue for these?

File: clr.defines.targets [hunk 2]
```diff
@@ -2,6 +2,7 @@
     <!-- Features we're currently flighting, but don't intend to ship in officially supported releases -->
     <PropertyGroup Condition="'$(IsPrerelease)' == 'true'">
```
Comment (jkotas) [inline clr.defines.targets]: Could you please make sure that this is not set in the Release branch? (It is not the case today.)

File: src/System.Private.CoreLib/src/System/Utf8String.Construction.cs [hunk 3]
```diff
@@ -0,0 +1,241 @@
...
+using System.Runtime.InteropServices;
+using System.Text;
+using System.Text.Unicode;
+
+namespace System
+{
+    public sealed partial class Utf8String
+    {
+        /*
+         * CONSTRUCTORS
```
Comment (jkotas) [inline src/System.Private.CoreLib/src/System/Utf8String.Construction.cs]: Assuming that this will move the shared partition eventually, this is not the best place to document this since it applies to a single runtime only. (The plumbing for Mono or CoreRT will be different.)

File: src/System.Private.CoreLib/shared/System/Text/Unicode/Utf8.cs (hunk 1 above)
Comment (GrabYourPitchforks) [reply src/System.Private.CoreLib/shared/System/Text/Unicode/Utf8.cs]: The `Utf8` static class is approved. See https://github.com/dotnet/corefx/issues/34761 and https://github.com/dotnet/apireviews/pull/89.

File: clr.defines.targets (hunk 2 above)
Comment (GrabYourPitchforks) [reply clr.defines.targets]: It is still _true_ in the _release/3.0_ branch currently, but that's because there's not yet an officially shipping release coming from that branch. This was set correctly to _false_ in the _release/2.x_ branches (see https://github.com/dotnet/coreclr/commit/5b5dca3254e968de60524621c10fec3fd5dc3bd0).

File: src/System.Private.CoreLib/shared/System/Text/Unicode/Utf8.cs (hunk 1 above)
Comment (jkotas) [reply src/System.Private.CoreLib/shared/System/Text/Unicode/Utf8.cs]: I would be better to submit that as separate PR from the experimental Utf8 string with appropriate tests, etc.

File: clr.defines.targets (hunk 2 above)
Comment (jkotas) [reply clr.defines.targets]: The previews shipping out of release/3.0 are proxies for officially shipping releases. We should flip that in release/3.0 now. It has not been flipped because of we do not have anything really depending on it for 3.0.

File: src/System.Private.CoreLib/src/System/Utf8String.Construction.cs (hunk 3 above)
Comment (GrabYourPitchforks) [reply src/System.Private.CoreLib/src/System/Utf8String.Construction.cs]: That's fair. Would it be ok to punt this until we do move it into the shared partition and have a better view of what ctor support might look like on those other platforms?

**Comment by GrabYourPitchforks** [comment]
@dotnet-bot test Windows_NT x64 Release CoreFX Tests

File: src/System.Private.CoreLib/src/System/Utf8String.Construction.cs (hunk 3 above)
Comment (jkotas) [reply src/System.Private.CoreLib/src/System/Utf8String.Construction.cs]: I would move this comment to somewhere in VM, e.g. ecall.cpp around PopulateManagedStringConstructors

Comment (jkotas) [reply src/System.Private.CoreLib/src/System/Utf8String.Construction.cs]: It applies equally to both String and Utf8String.

File: src/System.Private.CoreLib/shared/System/Text/Unicode/Utf8.cs [hunk 4]
```diff
@@ -0,0 +1,186 @@
...
+         * ill-formed subsequence can be deduced from these values.
+         *
+         * If 'replaceInvalidSequences' is true, the method is guaranteed never to return
+         * OperationStatus.InvalidData. If 'isFinalBlock' is true, the method is
+         * guaranteed never to return OperationStatus.NeedMoreData.
+         */
+
+        public static OperationStatus FromUtf16(ReadOnlySpan<char> source, Span<byte> destination, out int numCharsRead, out int numBytesWritten, bool replaceInvalidSequences = true, bool isFinalBlock = true)
+        {
+            // TODO: Replace this slow implementation with a bulk transcoder.
```
Comment (AaronRobinsonMSFT) [inline src/System.Private.CoreLib/shared/System/Text/Unicode/Utf8.cs]: Do we want to use the "TODO_UTF8STRING" for everything here? I see a mix in this review.

File: src/System.Private.CoreLib/shared/System/Text/Unicode/Utf8Utility.cs [hunk 5]
```diff
@@ -0,0 +1,99 @@
...
+        /// <summary>
+        /// Returns the byte index in <paramref name="utf8Data"/> where the first invalid UTF-8 sequence begins,
+        /// or -1 if the buffer contains no invalid sequences. Also outs the <paramref name="isAscii"/> parameter
+        /// stating whether all data observed (up to the first invalid sequence or the end of the buffer, whichever
+        /// comes first) is ASCII.
+        /// </summary>
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        public static int GetIndexOfFirstInvalidUtf8Sequence(ReadOnlySpan<byte> utf8Data, out bool isAscii)
+        {
+            // TODO: Make this slow implementation faster.
```
Comment (AaronRobinsonMSFT) [inline src/System.Private.CoreLib/shared/System/Text/Unicode/Utf8Utility.cs]: This kind of comment/TODO is less useful and/or noise since it is a general - "make things faster". The previous few mentioned a specific approach to get better perf, is there one here?

Comment (GrabYourPitchforks) [reply src/System.Private.CoreLib/shared/System/Text/Unicode/Utf8Utility.cs]: https://github.com/dotnet/coreclr/pull/21948 is an optimized drop-in replacement for this method. I should update the comment to call this out explicitly (and use `TODO_UTF8STRING`, as you suggested elsewhere).

File: src/System.Private.CoreLib/shared/System/Text/Unicode/Utf8.cs (hunk 1 above)
Comment (GrabYourPitchforks) [reply src/System.Private.CoreLib/shared/System/Text/Unicode/Utf8.cs]: Opened https://github.com/dotnet/coreclr/pull/23219 to track this.

**Comment by GrabYourPitchforks** [comment]
I reran the perf benchmarks for the `Memory<byte>.Span` property getter since that code path was modified. Results are below.

|                       Method |      Toolchain |     Mean |     Error |    StdDev | Ratio | RatioSD |
|----------------------------- |--------------- |---------:|----------:|----------:|------:|--------:|
|  GetSpan_MemOfBytesFromArray |    2.2-release | 4.244 ms | 0.0882 ms | 0.1474 ms |  1.00 |    0.00 |
|  GetSpan_MemOfBytesFromArray |     3.0-master | 1.933 ms | 0.0373 ms | 0.0366 ms |  0.45 |    0.02 |
|  GetSpan_MemOfBytesFromArray | dev-utf8string | 2.268 ms | 0.0280 ms | 0.0262 ms |  0.53 |    0.03 |
|                              |                |          |           |           |       |         |
| GetSpan_MemOfBytesFromMemMgr |    2.2-release | 5.306 ms | 0.0996 ms | 0.0978 ms |  1.00 |    0.00 |
| GetSpan_MemOfBytesFromMemMgr |     3.0-master | 3.150 ms | 0.0719 ms | 0.0769 ms |  0.59 |    0.02 |
| GetSpan_MemOfBytesFromMemMgr | dev-utf8string | 3.101 ms | 0.0270 ms | 0.0239 ms |  0.58 |    0.01 |
|                              |                |          |           |           |       |         |
|      GetSpan_MemOfBytesEmpty |    2.2-release | 1.393 ms | 0.0163 ms | 0.0153 ms |  1.00 |    0.00 |
|      GetSpan_MemOfBytesEmpty |     3.0-master | 1.108 ms | 0.0208 ms | 0.0240 ms |  0.79 |    0.02 |
|      GetSpan_MemOfBytesEmpty | dev-utf8string | 1.141 ms | 0.0227 ms | 0.0213 ms |  0.82 |    0.02 |

Overall the results are about what I'd expect. The `byte[]` case sees the largest regression because there's now an additional "are you `Utf8String`?" check where there was none before. However, due to the optimizations introduced earlier in the 3.0 timeframe (see https://github.com/dotnet/coreclr/pull/20386) we're still considerably better than 2.2. The `MemoryManager<byte>` case sees essentially no impact because that scenario was already expensive to begin with, and the extra few instructions are within the realm of noise.

File: src/System.Private.CoreLib/src/System/Char8.cs [hunk 6]
```diff
@@ -0,0 +1,69 @@
...
+    public readonly struct Char8 : IComparable<Char8>, IEquatable<Char8>
+    {
+        private readonly byte Value;
+
+        private Char8(byte value)
+        {
+            Value = value;
+        }
+
+        public static bool operator ==(Char8 a, Char8 b) => a.Value == b.Value;
```
Comment (justinvp) [inline src/System.Private.CoreLib/src/System/Char8.cs]: These should be named `left` and `right`, not `a` and `b`, according to the Framework Design Guidelines:

https://docs.microsoft.com/en-us/dotnet/standard/design-guidelines/naming-parameters

> **✓ DO** use `left` and `right` for binary operator overload parameter names if there is no meaning to the parameters.

File: src/System.Private.CoreLib/src/System/Char8.cs [hunk 7]
```diff
@@ -0,0 +1,69 @@
...
+// See the LICENSE file in the project root for more information.
+
+namespace System
+{
+    /// <summary>
+    /// Represents a UTF-8 code unit, the elemental type of <see cref="Utf8String"/>.
+    /// </summary>
+    public readonly struct Char8 : IComparable<Char8>, IEquatable<Char8>
+    {
+        private readonly byte Value;
```
Comment (justinvp) [inline src/System.Private.CoreLib/src/System/Char8.cs]: Why isn't this named `_value` according to the [coding style](https://github.com/dotnet/corefx/blob/master/Documentation/coding-guidelines/coding-style.md)?

> 3. We use `_camelCase` for internal and private fields and use readonly where possible. Prefix internal and private instance fields with `_` [...]

File: src/System.Private.CoreLib/src/System/Utf8String.cs [hunk 8]
```diff
@@ -0,0 +1,252 @@
...
+        private readonly byte _firstByte;
+
+        /*
+         * OPERATORS
+         */
+
+        /// <summary>
+        /// Compares two <see cref="Utf8String"/> instances for equality using a <see cref="StringComparison.Ordinal"/> comparer.
+        /// </summary>
+        public static bool operator ==(Utf8String a, Utf8String b) => Equals(a, b);
```
Comment (justinvp) [inline src/System.Private.CoreLib/src/System/Utf8String.cs]: Same comment as https://github.com/dotnet/coreclr/pull/23209/files#r264965988.

I realize `String`'s `==` operator uses `a` and `b`, but usually for new types, the guidelines are followed. For example, `DateTime` uses `d1` and `d2` as the parameter names, but when `DateTimeOffset` was added, the design guidelines were followed and `left` and `right` were used.

File: src/System.Private.CoreLib/src/System/Utf8String.Manipulation.cs [hunk 9]
```diff
@@ -0,0 +1,117 @@
...
+            if ((ulong)(uint)startIndex + (ulong)(uint)length > (ulong)(uint)this.Length)
+                ValidateStartIndexAndLength_Throw(startIndex, length);
+#else
+            if ((uint)startIndex > (uint)this.Length || (uint)length > (uint)(this.Length - startIndex))
+                ValidateStartIndexAndLength_Throw(startIndex, length);
+#endif
+        }
+
+        [StackTraceHidden]
+        [MethodImpl(MethodImplOptions.NoInlining)]
```
Comment (pentp) [inline src/System.Private.CoreLib/src/System/Utf8String.Manipulation.cs]: No-inline attribute's side-effects is that the method is no longer recognized by JIT as never returning and thus its callsites are not marked as cold code and are also emitted in a way that allows returning (registers preserved, reversed branch condition, etc).
Could be solved by not using `ThrowHelper` or by adding a `throw null;´ to the end, but no-inline should be removed either way.
The "proper/final" solution could be the never/bottom type: https://github.com/dotnet/csharplang/issues/538

File: src/System.Private.CoreLib/shared/System/Text/UnicodeUtility.cs [hunk 10]
```diff
@@ -11,7 +11,7 @@ internal static class UnicodeUtility
         /// <summary>
         /// The Unicode replacement character U+FFFD.
         /// </summary>
-        public const uint ReplacementChar = 0xFFFDU;
+        public const uint ReplacementChar = 0xFFFDu;
```
[the remark this answers is not in the document]
Comment (GrabYourPitchforks) [reply src/System.Private.CoreLib/shared/System/Text/UnicodeUtility.cs]: Isn't there some strange corner case where the C# compiler might ignore the declared type of the `const` field and use the literal value instead, which could end up inadvertently making this an `int` or similar depending on the call site? I may be remembering the language rules incorrectly.

File: src/System.Private.CoreLib/src/System/Utf8Extensions.cs [hunk 11]
```diff
@@ -0,0 +1,367 @@
...
+        /// <remarks>Returns default when <paramref name="text"/> is null.</remarks>
+        /// <exception cref="System.ArgumentOutOfRangeException">
+        /// Thrown when the specified <paramref name="start"/> index or <paramref name="length"/> is not in range.
+        /// </exception>
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        public static ReadOnlySpan<Char8> AsSpan(this Utf8String text, int start, int length)
+        {
+            if (text == null)
+            {
+                if (start != 0 || length != 0)
```
[the remark this answers is not in the document]
Comment (GrabYourPitchforks) [reply src/System.Private.CoreLib/src/System/Utf8Extensions.cs]: Good eye! This was copied from `MemoryExtensions`, but that same optimization can be made there. Even better, if https://github.com/dotnet/coreclr/issues/19074 is fixed, then you don't even need the extra `test` instructions, and the entire code path now becomes:

```asm
test text, text
jnz NOT_NULL
or start, length
jne THROW_ARGEXCEPTION
; <return default >
```

(The JIT should be smart enough to realize that _start_ and _length_ are no longer needed once this branch starts running, so it's free to corrupt their values while performing the arg checks.)

File: src/System.Private.CoreLib/shared/System/Text/UnicodeUtility.cs (hunk 10 above)
[the remark this answers is not in the document]
Comment (jkotas) [reply src/System.Private.CoreLib/shared/System/Text/UnicodeUtility.cs]: That would sound like a compiler bug to me....

[the remark this answers is not in the document]
Comment (GrabYourPitchforks) [reply src/System.Private.CoreLib/shared/System/Text/UnicodeUtility.cs]: I've asked some compiler folks to respond when they have a sec. It's likely I'm just misremembering or being overly paranoid.

[the remark this answers is not in the document]
Comment (agocke) [reply src/System.Private.CoreLib/shared/System/Text/UnicodeUtility.cs]: The type of the literal is technically an int, but it has a constant value that is in the range of `uint`, so there is an implicit constant expression conversion to `uint`. If the conversion were unsafe it would be an error.

File: src/System.Private.CoreLib/src/System/Utf8Extensions.cs (hunk 11 above)
[the remark this answers is not in the document]
Comment (gfoidl) [reply src/System.Private.CoreLib/src/System/Utf8Extensions.cs]: > that same optimization can be made there

https://github.com/dotnet/coreclr/pull/23239

File: src/System.Private.CoreLib/shared/System/Text/Unicode/Utf8.cs (hunk 4 above)
Comment (GrabYourPitchforks) [reply src/System.Private.CoreLib/shared/System/Text/Unicode/Utf8.cs]: Yes. I'll normalize all of the TODOs and give them more specific comments than simply "improve perf". All of the TODOs should direct us to consider very specific courses of action.

File: src/System.Private.CoreLib/src/System/Utf8String.Manipulation.cs (hunk 9 above)
Comment (pentp) [reply src/System.Private.CoreLib/src/System/Utf8String.Manipulation.cs]: Should still add a `throw null` at the end of the method so that is properly detected as never returning (JIT detects it by observing no `ret` instructions in the method body).
Compare [Substring w/o and w/ throw null](https://sharplab.io/#v2:EYLgHgbALANALiAhgZwLYB8ACAGABJgRgG4BYAKB3wIDoAlAVwDs4BLVAU2oGEB7VABxYAbdgCcAymIBuLAMbtkpMuUwBmfACZcAQVwBvcriO4WzXABl2jAOZwAFkuO5Dxtbh7AAVu1lxc4+mBCbAAKUz9kOERROABJRgATdjAYEzMRG3sAShcjAzInJwA1RCEWBMQ4dnEomPiksG1Ey0y7EMjouMTk1IzbOyylXMKAehHcAHl+VlQWAC9Klh5GZBBcZFN5NI7GeWRcaPYTVFR6KOARVIB3I9lERnwAdlwAcnsWZBf3UVx7I4BrRg8K6MYZOMa4ACiAjgAE9tlFdkcWAAzX52W6lEQ/fgoZDsBK4ej7RDrWpdBoHPx/Kh4YA8JgVUSw6jkMHGVG4EJ9ey4AC8fNw2ByBUKeXZYswzwARNLHGLcABfNmisWc7lWfr8wXvZDUFr9EUK/QSwpS9EfeVi5XKVVm57xKqiRilAJBAihDp1bopXA8gZWpUqpwAbQAsux7DwErEBEIQhGozG41NWMs9QA5HjxMqMUzWLIAXWG/FELCklSOukdYhdQjdwTCZi9FJ6aT8/qy/IAfLhGPQhEIhnbw5G7NHY/x44nx8mp6mlitqNprNZRAoNlJ2DnTPmiyWyxWqvgoLgSmUKlUap16skmgkDfYmxFybffeE/ZrssN8saIZJbj4DgzFMfxcUYAAeAAVbtqHEMothRHgfnHK4LX2K5kP+PVTXVEIB2WAt8PCLIWzfXAAGouQImwsmI5gsn9XBe3woRCLo+gSN1fUvwGU1ilKcpK2vb0GnvR87AAfSguxRGBdpXx9XpeMGYYbQPctKxPM9BMvapFLE5peOk2T5I/MilPbT9WiNYxfwVPDOIYizKRYpy4CybiJNssV7ONIwZLkq5tFEax6GAuAJjOCYUVoe5rHYSEwHkaZFxCR5VLtJwbWNdghHxU0/P8wLgRCsKIqiyLYvixLkvYVLlhCAAOTLjRy4x1LIABIUtNOPQgIG0krgtC8KrEi6LqpsWqUrTRhn1wKQuz5Xt7CCvsByHchOrcTAtAAIRNO0PwkwNhjcDxvF8fxAkbcyDLbD9Ox/U1zyEq8HsaIzWgUm9LM7Yc/3GBdZgWObVnWTZkRWRE9gOddjlOc5LlwG5cDuB5zTeOwPi+ZD0QBIEQVNCFoWmeFTB2LZORpO5BzEXBcWQfFCWJA4yT+ylKgJ2lcHpRlohZYMHLRDVWm1IUfMKIrjXNWVA2y4W1VFpiBXQnibMK/jXGeXUFY6pX7VwGtnVdW6PV+0S2wBtTDdHJNJ2nMcJxTBqlyzHc81o4s7V6o8jkOk26wbC37s5x70hUnsNsHQHjHt2dHYTZ25yEBd02XVd12Z8tt0YXM9x9pw/a0zBTzevSRNbL6H14haXIjjsVJerLjH/dhAJOca0jA+5oNg+C5COJCUOBdDUawnDW6MRy2No+iPIbsBKOoueiPcxjeOY1f2IXzycb1bztaMCvhM+8TjOGy3q+UmzA064vD1L8vdLP8Oa4kkygvrz7Uie5u7Qyw5KLPeS9t57y8gA40QCxTDTKmNZglUYpxWmklWaaUMr6yMO1MUeUCrTyOv5YwcDRoVUmighKaD6pzWaq1BUODChrTHv2WOttbSPz6kcAaQ1TIjXKuNJBU1KF1TdgtJa0cmFoRYVtMgiogA).

Comment (jkotas) [reply src/System.Private.CoreLib/src/System/Utf8String.Manipulation.cs]: Or just use regular `throw` without the throw helpers goop. It will solve all the problems discussed here.

File: src/System.Private.CoreLib/src/System/Utf8Extensions.cs (hunk 11 above)
[the remark this answers is not in the document]
Comment (GrabYourPitchforks) [reply src/System.Private.CoreLib/src/System/Utf8Extensions.cs]: I reverted this suggestion because per the newly opened issue, it looks like RyuJIT already handles this for us.

**Comment by GrabYourPitchforks** [comment]
(In the latest iteration I merged master manually because GitHub was having trouble auto-resolving merge conflicts.)

File: clr.defines.targets (hunk 2 above)
Comment (GrabYourPitchforks) [reply clr.defines.targets]: I also kicked off a local build with `IsPrerelease=false` and ensured that no part of the feature inadvertently makes its way in to System.Private.CoreLib.dll.

**Comment by yahorsi** [comment]
May I ask silly question, are there any perf tests showing benefits of having special Utf8String string?

**Comment by GrabYourPitchforks** [comment]
@yahorsi - Not a silly question at all. :) I have some rough numbers showing this, but they're from an earlier iteration of this prototype. Off the top of my head here are the quick vitals:

- Given a binary buffer that represents UTF-8 data, creating a `Utf8String` instance from that buffer is approx. 5x faster than creating a `String` instance.

- For most payloads, UTF-8 consumes less memory than UTF-16. Our data show that most strings still consist of mainly ASCII data, even in programs used primarily by East Asian users. (East Asian is particularly interesting here because UTF-8 representation tends to expand it, not compact it, when compared to UTF-16.) Assuming you could flow `Utf8String` instances instead of `String` instances throughout your applications, you should expect to see a double-digit percentage decrease in overall memory utilization.

- `Utf8String` is slightly more compact than using `byte[]` for the same data. It's only 4 bytes or so on average, but these add up quickly, especially when that represents 10 - 15% of the total object size of a small object.

- Projecting from `Utf8String` back to `ROS<byte>` or `ROM<byte>` so that the data can be sent across the wire is an O(1) operation, compared to an O(n) operation for regular `String`.

- Given ASCII `Utf8String` instances, certain text operations (`ToUpperInvariant` and friends) perform at around 2x the throughput of those same operations on `String`. This makes sense: they contain the same amount of information in half the memory.

There are also some drawbacks.

- `ToUpperInvariant` on non-ASCII inputs (or `ToUpper` in locales like _tr-TR_, even over ASCII input) are much more expensive on `Utf8String` than they are on `String.` This is because the localization tables use UTF-16 representation natively, and calling into them requires temporarily transcoding the input data.

- If you find yourself switching between `Utf8String` and `String` frequently, there will be considerable overhead to changing the representation. The best course of action is to stick with one representation for as long as you can, only switching between the two if / when absolutely necessary. This could make the type a bit rough to use until there's a healthy ecosystem of APIs built up around it. (And most APIs aren't going to be enlightened about this type since there's no real need to do so.)

In short, the best performance and memory utilization is going to be seen when reading data from the wire, keeping it in UTF-8 representation (instead of converting to UTF-16), performing operations on it while remaining in the UTF-8 world, then sending it back out again to wherever it needs to go.

**Comment by yahorsi** [comment]
Am I right ASP.NET/Kestrel guys didn't try to utilise it yet?

**Comment by benaadams** [comment]
> Am I right ASP.NET/Kestrel guys didn't try to utilise it yet?

Its not been added yet so ASP.NET/Kestrel can't currently use it

**Comment by GrabYourPitchforks** [comment]
@dotnet-bot test this please

**Comment by jkotas** [comment]
> ASP.NET/Kestrel

It is unclear whether this type would help Kestrel much. If you are just using UTF8 as an internal implementation detail, you can get all the performance benefits by simply using byte arrays and Spans.

We do not have data and agreement that it is a good idea to add this type. New string exchange type comes with extremely large cost to the ecosystem. It is why this is added as experimental feature under ifdef that is not going to ship in 3.0.

We do have an agreement that it is a good idea to add various Utf8 helpers that operate on Spans and support using UTF8 as an internal implementation detail. Those helpers are gradually being added.

**Comment by yahorsi** [comment]
If we think about applications of the Utf8 there are a lot in the WEB world. But, currently, ASP.NET Core is  defined using strings everywhere in public interfaces and that cannot be changed. So my bet is that unfortunately we won't be able to get enough benefits as in the final end we'll have to convert Utf8String anyway to just String. So, do you think it make sense to run another research to have same type String but have an option to still store Utf8 internally? Compiler could do something when he sees fixed & string, like call something to convert to utf16 first, or even additionally.

**Comment by jkotas** [comment]
> ASP.NET Core is defined using strings everywhere in public interfaces

It is not just ASP.NET Core. It is *everything* in .NET.

> it make sense to run another research to have same type String but have an option to still store Utf8 internally

https://github.com/dotnet/coreclr/issues/7083 has discussions about potential designs and tradeoffs. It is a ton of complex work to make something like this work and the estimated returns are not that big (can get better returns on investment elsewhere)...

**Comment by GrabYourPitchforks** [comment]
CI was clogged due to some outages from yesterday. I'm hoping that the latest iteration will make it through unscathed.

**Comment by GrabYourPitchforks** [comment]
@yahorsi This is an area of active investigation runtime-wide and Framework-wide. The `Utf8String` type is one potential experiment, and it's possible this won't pan out. As @jkotas said this is all being done under an ifdef so that it can be ripped out easily if necessary.

There are discussions about the merits of the various proposals in existing tracking issues (see https://github.com/dotnet/coreclr/issues/7083, https://github.com/dotnet/corefx/issues/34094, https://github.com/dotnet/corefx/issues/30503, https://github.com/dotnet/corefxlab/issues/2350, and https://github.com/dotnet/corefxlab/issues/2368 for various discussions). It would help keep things more focused and organized if we reuse those existing issues for feature discussion and instead limit the conversation here on the actual code under review.

**Comment by benaadams** [comment]
Will `Utf8String` have value equality with `string`? e.g.
```csharp
public static bool operator ==(Utf8String left, string right) 
public static bool operator ==(string left, Utf8String right) 
```
As string and Utf8String have known encodings; so should be able to be compared without allocating the other type to change the encoding to compare.

> We do have an agreement that it is a good idea to add various Utf8 helpers that operate on Spans ...

Similarly then (not sure name)
```csharp
ValueEquals(this ReadOnlySpan<char>, ReadOnlySpan<Char8>)
ValueEquals(this ReadOnlySpan<Char8>, ReadOnlySpan<char>)
```

File: src/System.Private.CoreLib/src/System/Utf8String.Searching.cs [hunk 12]
```diff
@@ -0,0 +1,93 @@
...
+        }
+
+        // Ordinal search
+        public bool Contains(Rune value)
+        {
+            // TODO_UTF8STRING: This should be split into two methods:
+            // One which operates on a single-byte (ASCII) search value,
+            // the other which operates on a multi-byte (non-ASCII) search value.
+
+            Span<byte> runeBytes = stackalloc byte[Utf8Utility.MaxBytesPerScalar];
```
[the remark this answers is not in the document]
Comment (GrabYourPitchforks) [reply src/System.Private.CoreLib/src/System/Utf8String.Searching.cs]: @gfoidl I was experimenting in a separate branch with moving some of this into the JIT if possible, using a technique very similar to what you had done. But it'll be a while before I get around to performing these sorts of optimizations. Pseudocode:

```cs
public bool Contains(Rune value)
{
    if (JitHelpers.TryConstRuneToAscii(value, out byte valueAscii)
    {
        // if the JIT knows that 'value' is const and ASCII, go down this code path
        // (valueAscii would've gotten baked into the codegen as a constant)
        return SpanHelpers.IndexOf(..., valueAscii) >= 0
    }
    else if (JitHelpers.TryConstRuneToUtf8(value, out int data, out int dataLength))
    {
        // if the JIT knows that 'value' is const and not ASCII, go down this code path
        // (data and dataLength would've gotten baked into the codegen as constants)
        return SpanHelpers.IndexOf(..., ref Unsafe.As<int, byte>(ref data), dataLength) >= 0;
    }
    else
    {
        // if the JIT can't reason about the const-ness of 'value', go down this code path
        // stackalloc and convert at runtime
    }
}
```

[the remark this answers is not in the document]
Comment (GrabYourPitchforks) [reply src/System.Private.CoreLib/src/System/Utf8String.Searching.cs]: (In the above example it would be perfectly valid for the `TryConstXyz` methods to be hardcoded to always return `false`, and the first two branches would be folded away by the JIT. The end result is that the code would still work; it just wouldn't be as optimized.)

**Comment by GrabYourPitchforks** [comment]
CI still on the floor?
@dotnet-bot test OSX10.12 x64 Checked Innerloop Build and Test

**Comment by GrabYourPitchforks** [comment]
@dotnet-bot test Windows_NT x64 Checked gcstress0x3
@dotnet-bot test Windows_NT x64 Checked gcstress0xc_minopts_heapverify1
@dotnet-bot test Windows_NT x64 Checked gcstress0xc_jitstress1
@dotnet-bot test Windows_NT x64 Checked gcstress0xc_jitstress2
@dotnet-bot test Windows_NT x64 Checked gcstress0xc_zapdisable_heapverify1
@dotnet-bot test Windows_NT x64 Checked gcstress0xc_zapdisable_jitstress2
@dotnet-bot test Windows_NT x64 Checked gcstress0xc_zapdisable
@dotnet-bot test Windows_NT x64 Checked gcstress0xc

**Comment by GrabYourPitchforks** [comment]
@dotnet-bot test this please
